### ISSUE TITLE
jobs and backgrounding broken

### ISSUE THREAD

**Issue by laerus** [xonsh/xonsh#2071]
```
$ vim 
( hitting Ctrl-Z)

$                                     
$                        
$ fg
[1]+ running: vim (1966)
$ bg
[1]+ running: vim (1966)
$ fg
$ ( hitting Ctrl-D)
xonsh: there is an unfinished job
-----
[1]+ running: vim (1966)
-----
Type "exit" or press "ctrl-d" again to force quit.
```

same happens with `top`

**Comment by scopatz** [comment]
Thanks for reporting @laerus! I am not sure what is going on here.

**Comment by scopatz** [comment]
Though I am seeing the same issues.

**Comment by laerus** [comment]
been trying to troubleshoot this from time to time with no luck,
this was an issue before the subproc refactor

**Comment by adqm** [comment]
I'm not sure when this became an issue.  In case it helps, I'm running a version forked from 0.3.4 and I don't have this issue.  That said, probably a lot has changed with job control since then.

**Comment by scopatz** [comment]
Thanks @adqm. If someone feels like bisecting, that could be helpful to know when it broke :)

**Comment by gforsyth** [comment]
I think there are two separate issues here (that I've noticed). 
One is that backgrounding doesn't seem to work (the big issue). The other is that while the SIGSTOP is sent by Ctrl-Z, the prompt doesn't always return control to the user. 

Bisecting and running `vim` and then hitting `Ctrl-Z` points to 008085e as the culprit for control not being returned to the user. And commenting out https://github.com/xonsh/xonsh/blob/master/xonsh/proc.py#L1739 does fix that issue (although it may introduce other problems). 

Still working on tracking down exactly where bringing vim back to the foreground broke...

**Comment by scopatz** [comment]
Thanks for hunting this down @gforsyth! I am formulating an idea on how to fix this.  I think the proc.wait() might only be needed in we are running an alias (Proxy or ProxyThread). Or maybe this needs to be `proc.communicate()` if we have a Popen class.

**Comment by scopatz** [comment]
Ok, communicate didn't work.

**Comment by scopatz** [comment]
Alright, so I think part of the issue might be whether we want to call wait() and end() depend on what is returned from `wait_for_active_job()`. Unfortunately right now, Crtl-z and a process exiting normally will both return None. So, I think that job control needs to be updated to reflect the kind of signal it sees before we can fix the iterraw() implementation.

**Comment by scopatz** [comment]
Was this fixed by the recent merge to master or are there still lingering issues here?

**Comment by laerus** [comment]
yea this works for both vim and top, thank you @scopatz !

**Comment by laerus** [comment]
something is going on with the python interp when trying to background it
```
$ python3
Python 3.5.2+ (default, Dec 13 2016, 14:16:35) 
[GCC 6.2.1 20161124] on linux
Type "help", "copyright", "credits" or "license" for more information.
$ 
$ 
$ jobs
[1]+ running: python3 (12032)
$ fg
[1]+ running: python3 (12032)

^C <------after Ctrl-C here i pressed Ctrl-D to make it throw the exception

xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1311, in run
    r = self.f(self.args, sp_stdin, sp_stdout, sp_stderr, spec)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1151, in proxy_two
    return f(args, stdin)
  File "/home/laerus/src/xonsh/xonsh/jobs.py", line 375, in fg
    wait_for_active_job()
  File "/home/laerus/src/xonsh/xonsh/jobs.py", line 192, in wait_for_active_job
    obj.returncode = os.WEXITSTATUS(wcode)
AttributeError: can't set attribute
$ 
$ jobs
[1]+ running: python3 (12032)
$ fg
[1]+ running: python3 (12032)
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1311, in run
    r = self.f(self.args, sp_stdin, sp_stdout, sp_stderr, spec)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1151, in proxy_two
    return f(args, stdin)
  File "/home/laerus/src/xonsh/xonsh/jobs.py", line 375, in fg
    wait_for_active_job()
  File "/home/laerus/src/xonsh/xonsh/jobs.py", line 183, in wait_for_active_job
    _, wcode = os.waitpid(obj.pid, os.WUNTRACED)
ChildProcessError: [Errno 10] No child processes
```

**Comment by scopatz** [comment]
This is probably related to #1368 and #1478

**Comment by scopatz** [comment]
So I think there might be two issues:

1. interactive Python is being run as a threadable command, which is messing things up.
2. After background we are not turning echo back on.

**Comment by scopatz** [comment]
Ok! I think #2123 might fix this.  It works for me on ubuntu on readline and ptk

**Comment by laerus** [comment]
@scopatz i've tried the snippet from #830 and got:
```
$ bash -c 'while true; do echo hey; sleep 1; done'
hey
hey
^C
^C^C^C

^C^C^C^C

^C^C^C
^C^C^C^C^C^C^C^C^C^C^C

xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/laerus/src/xonsh/xonsh/base_shell.py", line 322, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/laerus/src/xonsh/xonsh/codecache.py", line 65, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/home/laerus/src/xonsh/xonsh/built_ins.py", line 858, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/home/laerus/src/xonsh/xonsh/built_ins.py", line 826, in run_subproc
    command.end()
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1908, in end
    for _ in self.tee_stdout():
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1833, in tee_stdout
    for line in self.iterraw():
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 1796, in iterraw
    time.sleep(timeout * cnt)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 751, in _signal_int
    self._restore_sigint(frame=frame)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 764, in _restore_sigint
    old(signal.SIGINT, frame)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  ...
  signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 749, in _signal_int
    self.send_signal(signum)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 882, in send_signal
    time.sleep(1e-7)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 749, in _signal_int
    self.send_signal(signum)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 882, in send_signal
    time.sleep(1e-7)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
...
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 753, in _signal_int
    signal.pthread_kill(threading.get_ident(), signal.SIGINT)
  File "/home/laerus/src/xonsh/xonsh/proc.py", line 752, in _signal_int
    if on_main_thread():
  File "/home/laerus/src/xonsh/xonsh/tools.py", line 751, in on_main_thread
    return threading.current_thread() is threading.main_thread()
  File "/usr/lib/python3.5/threading.py", line 1232, in current_thread
    return _active[get_ident()]
RecursionError: maximum recursion depth exceeded in comparison
```

**Comment by laerus** [comment]
not related to backgrounding but this didn't happen before #2123

**Comment by scopatz** [comment]
That snippet seems to work for me.  And are you sure that you have #2087 merged in?

**Comment by stan3** [comment]
With ctrl-z in 0.5.9 on linux things like vim and top work but with sleep the process is suspended but control is not return to the prompt.

**Comment by mitnk** [comment]
@stan3 Could you open a new issue describing the details how to reproduce this problem? (e.g. what do you mean by "with sleep"). Thanks.

**Comment by stan3** [comment]
sleep as in:

```
~  sleep 30
^Z
fg
```

where ~ is the shell prompt

**Comment by leonqli** [comment]
ctrl-z doesn't work....

%sleep 10
^Z^Z^Z^Z

xonsh/0.6.7
